### PR TITLE
Tidy land-detected handling

### PR DESCRIPTION
**Pull request by peterbarker**
This was calling several methods on the mode to work out what that mode wants to do when the landing detector fires.

The logic should be the same here, just unrolled to be legible.

I've added a test for the parameter used to change the vehicle behaviour.

### CODE REVIEW

**Comment by peterbarker** [comment]
```
Board                    AP_Periph  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
CubeOrange-periph-heavy  *                                                                     
Durandal                            *      *           -16     -16               *      *      *
Hitec-Airspeed           *                                                                     
KakuteH7-bdshot                     *      *           0       0                 *      *      *
MatekF405                           *      *           0       0                 *      *      *
Pixhawk1-1M-bdshot                  *                  0       0                 *      *      *
f103-QiotekPeriph        *                                                                     
f303-Universal           *                                                                     
iomcu                                                                *                         
revo-mini                           *      *           0       0                 *      *      *
skyviper-v2450                                         0                                       
```

File: ArduCopter/mode.cpp [hunk 1]
```diff
@@ -472,6 +472,31 @@ void Copter::notify_flight_mode() {
...
+
+    if (has_manual_throttle()) {
+        return;
+    }
+    // we should probably have a "allows_disarm(...)" callback
+    if (!allows_arming(AP_Arming::Method::LANDING)) {
+        return;
+    }
+
+    AP::arming().disarm(AP_Arming::Method::LANDED);
```
Comment (peterbarker) [inline ArduCopter/mode.cpp:497]: How did we end up with both `LANDED` and `LANDING`?  Pre-existing, but we should clarify.

Perhaps `s/LANDING/LAND_DETECTOR` if it makes sense.

File: ArduCopter/land_detector.cpp [hunk 2]
```diff
@@ -156,13 +156,9 @@ void Copter::set_land_complete(bool b)
...
     set_likely_flying(!b);
-    
-    // trigger disarm-on-land if configured
-    bool disarm_on_land_configured = (g.throttle_behavior & THR_BEHAVE_DISARM_ON_LAND_DETECT) != 0;
-    const bool mode_disarms_on_land = flightmode->allows_arming(AP_Arming::Method::LANDING) && !flightmode->has_manual_throttle();
 
-    if (ap.land_complete && motors->armed() && disarm_on_land_configured && mode_disarms_on_land) {
-        arming.disarm(AP_Arming::Method::LANDED);
+    if (b) {
+        flightmode->land_complete();
```
Comment (peterbarker) [inline ArduCopter/land_detector.cpp:161]:
```suggestion
        flightmode->handle_land_complete();
```

File: ArduCopter/land_detector.cpp [hunk 3]
```diff
@@ -156,13 +156,9 @@ void Copter::set_land_complete(bool b)
...
     }
```
Comment (IamPete1) [inline ArduCopter/land_detector.cpp:162]:
```suggestion
    if (!b) {
        // Not landed
        return;
    }
    
    // trigger disarm-on-land if configured
    if ((g.throttle_behavior & THR_BEHAVE_DISARM_ON_LAND_DETECT) == 0) {
        // not configured to disarm
        return;
    }

    if (!motors->armed()) {
        // not armed
        return;
    }

    if (has_manual_throttle()) {
        return;
    }
    // we should probably have a "allows_disarm(...)" callback
    if (!allows_arming(AP_Arming::Method::LANDING)) {
        return;
    }

    AP::arming().disarm(AP_Arming::Method::LANDED);
```

File: ArduCopter/land_detector.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/land_detector.cpp:161]: Went the other way and rewrote in place.

File: ArduCopter/land_detector.cpp (hunk 3 above)
Comment (peterbarker) [reply ArduCopter/land_detector.cpp]: Yep, rewrote the thing in-place instead.

**Review by IamPete1 — APPROVED** [review]
LGTM

**Review by rmackay9 — APPROVED** [review]

**Pull request merged**